cmd/gitcloner/gogit/cloner.go

```diff
@@ -0,0 +1,174 @@
+	if err != nil {
+		return err
+	}
+
+	if opts.Branch == "" && opts.Revison == "" {
+		opts.Branch = defaultBranch
+		return cloneBranch(opts, auth, caBundle)
+	}
+
+	if opts.Branch != "" {
+		return cloneBranch(opts, auth, caBundle)
+	}
+
+	return cloneRevision(opts, auth, caBundle)
+}
+
+func cloneBranch(opts *cmd.Options, auth transport.AuthMethod, caBundle []byte) error {
+	_, err := plainClone(opts.Path, false, &git.CloneOptions{
+		URL:             opts.Repo,
+		Auth:            auth,
+		InsecureSkipTLS: opts.InsecureSkipTLS,
+		CABundle:        caBundle,
+		SingleBranch:    true,
+		ReferenceName:   plumbing.ReferenceName(opts.Branch),
+	})
+
+	return err
+}
+
+func cloneRevision(opts *cmd.Options, auth transport.AuthMethod, caBundle []byte) error {
+	r, err := plainClone(opts.Path, false, &git.CloneOptions{
+		URL:             opts.Repo,
+		Auth:            auth,
+		InsecureSkipTLS: opts.InsecureSkipTLS,
+		CABundle:        caBundle,
+	})
+	if err != nil {
+		return err
+	}
+	h, err := r.ResolveRevision(plumbing.Revision(opts.Revison))
+	if err != nil {
+		return err
+	}
+	w, err := r.Worktree()
+	if err != nil {
+		return err
+	}
+
+	return w.Checkout(&git.CheckoutOptions{
+		Hash: *h,
+	})
+}
+
+func getCABundleFromFile(path string) ([]byte, error) {
+	if path != "" {
+		caBundle, err := readFile(path)
+		if err != nil {
+			return nil, err
+		}
+
+		return caBundle, nil
+	}
+
+	return nil, nil
+}
+
+// addGitAuthToOpts adds auth for cloning git repos based on the parameters provided in opts.
+func createAuthFromOpts(opts *cmd.Options) (transport.AuthMethod, error) {
+	if opts.SSHPrivateKeyFile != "" {
+		privateKey, err := readFile(opts.SSHPrivateKeyFile)
+		if err != nil {
+			return nil, err
+		}
+		gitURL, err := giturls.Parse(opts.Repo)
+		if err != nil {
+			return nil, err
+		}
+		auth, err := gossh.NewPublicKeys(gitURL.User.Username(), privateKey, "")
+		if err != nil {
+			return nil, err
+		}
+		if opts.KnownHostsFile != "" {
+			knownHosts, err := readFile(opts.KnownHostsFile)
+			if err != nil {
+				return nil, err
+			}
+			knownHostsCallBack, err := createKnownHostsCallBack(knownHosts)
+			if err != nil {
+				return nil, err
+			}
+			auth.HostKeyCallback = knownHostsCallBack
+		} else {
+			//nolint G106: Use of ssh InsecureIgnoreHostKey should be audited
+			//this will run in an init-container, so there is no persistence
+			auth.HostKeyCallback = ssh.InsecureIgnoreHostKey()
+		}
+		return auth, nil
+	}
+
+	if opts.Username != "" && opts.PasswordFile != "" {
+		password, err := readFile(opts.PasswordFile)
+		if err != nil {
+			return nil, err
+		}
+
+		return &httpgit.BasicAuth{
+			Username: opts.Username,
+			Password: string(password),
+		}, nil
+	}
+
+	return nil, nil
+}
+
+func createKnownHostsCallBack(knownHosts []byte) (ssh.HostKeyCallback, error) {
+	f, err := os.CreateTemp("", "known_hosts")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(f.Name())
```

Comment:
Why deleting it? Is it loaded in memory?

Reply:
we need to create a temp file to pass it to `gossh.NewKnownHostsCallback`. We don't need it after the knownHosts callback has been created, that's why it is deleted